src/ray/common/asio/io_context.cc

```diff
@@ -0,0 +1,257 @@
+// Copyright 2017 The Ray Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "ray/common/asio/io_context.h"
+#include <algorithm>
+#include <cmath>
+#include <iomanip>
+#include <iostream>
+#include <utility>
+
+/// C++11 helper for generalized lambda capture.
+template <typename T, typename F>
+class capture_impl {
+  T x;
+  F f;
+
+ public:
+  capture_impl(T &&x, F &&f) : x{std::forward<T>(x)}, f{std::forward<F>(f)} {}
+
+  template <typename... Ts>
+  auto operator()(Ts &&... args) -> decltype(f(x, std::forward<Ts>(args)...)) {
+    return f(x, std::forward<Ts>(args)...);
+  }
+
+  template <typename... Ts>
+  auto operator()(Ts &&... args) const -> decltype(f(x, std::forward<Ts>(args)...)) {
+    return f(x, std::forward<Ts>(args)...);
+  }
+};
+
+template <typename T, typename F>
+capture_impl<T, F> capture(T &&x, F &&f) {
+  return capture_impl<T, F>(std::forward<T>(x), std::forward<F>(f));
+}
+
+/// A helper for creating a snapshot view of the stats for a handler.
+/// This acquires a reader lock on the provided guarded handler stats, and creates a
+/// lockless copy of the stats.
+HandlerStats to_stats_view(const GuardedHandlerStats &stats) {
+  absl::ReaderMutexLock lock(&(stats.mutex));
+  return HandlerStats(stats.stats);
+}
+
+/// Set stats collection flag based on corresponding Ray config flag.
+const bool io_context_proxy::stats_collection_enabled_ =
+    RayConfig::instance().asio_event_loop_stats_collection_enabled();
```

Comment:
Btw, this should happen after the config is initialized at each component. Otherwise, the won't be reflected. https://github.com/ray-project/ray/blob/2da58bb0214644177c3f9a9c59d03cc5fa0d1c47/src/ray/core_worker/core_worker.cc#L139 (I am mentioning just to make sure you didn't miss it)

Reply:
That's a great point, and also applies to the [raylet code path](https://github.com/ray-project/ray/blob/2da58bb0214644177c3f9a9c59d03cc5fa0d1c47/src/ray/raylet/main.cc#L120). And there it's even worse, since [we instantiate the `io_context` before we initialize the config](https://github.com/ray-project/ray/blob/2da58bb0214644177c3f9a9c59d03cc5fa0d1c47/src/ray/raylet/main.cc#L99-L100)... I think that we're going to have to check the config flag on every post, unfortunately.

Reply:
Given that the config instance get, and the config flag get, should both be very fast (the latter is inlined), I think we'll be ok with checking the flag on every post.

If it actually was an expensive operation, ideally there would be an `is_initialized_` flag on `RayConfig` that's set after `RayConfig::instance()::initialize()` is called, along with an `is_initialized()` member function. Then, at the beginning of `post()` in `io_context_proxy`, we could do the following check:
```cpp
if (!config_initialized_) {
  config_initialized_ = RayConfig::instance().is_initialized();
  if (config_initialized_) {
    stats_collection_enabled_ = RayConfig::instance().asio_event_loop_stats_collection_enabled();
  }
}
if (!stats_collection_enabled_) {
  // Config is not initialized or stats collection is not enabled, either way, don't collect stats.
  return boost::asio::io_context::post(std::move(handler));
}
```
That way, once the config is initialized, each if statement should be easy to predict and result in hot dynamic branch prediction, and should be close to the same overhead as if we had set `stats_collection_enabled_` on instantiation, for the rest of `io_context_proxy`'s lifetime.

But again, I think `RayConfig::instance().asio_event_loop_stats_collection_enabled()` is cheap enough to call on each `post()`, so we should just do that. Do you agree?